Added LcmPublisherSystem Based on System 2.0 Framework

liangfok opened the pull request:
This builds upon #2801 by adding a system that takes as input a `VectorInterface` and publishes LCM messages of type `drake::lcmt_drake_signal`.

Resolves #2696.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2821)
<!-- Reviewable:end -->

liangfok commented:
+@jwnimmer-tri for feature review (perhaps after #2801 is merged).

---

Review status: 0 of 22 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KMkO4NiBvbpLuv6chm-:123583090)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2743/ and is thus ready for feature review once #2801 is merged.

jwnimmer-tri commented:
Should not review until it's rebased (or merged) onto #2801 -- or master once if 2801 has merged.

liangfok commented:
@jwnimmer-tri: I've updated this PR based on the latest from master. I believe it is ready for feature review.

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KMzRVVVzCI2axrzlEj9:360392416)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/memory_buffer_utils.h, line 24 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KMzVPsI8d2kaXwdLeh3:-KMzVPsJXMZKXBfUVDS2:1846804474) ([raw file](https://github.com/robotlocomotion/drake/blob/638580f04e57c40658ac51081416ed2a79a8c0d2/drake/common/memory_buffer_utils.h#L24)):*
> MemoryBufferToString

AFAICT this is only ever called from the unit test.  Why is it in this PR?

It is difficult to evaluate its design without context about how it will be used -- is there a dev branch where its use is demonstrated?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/memory_buffer_utils.h, line 24 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KMzVPsI8d2kaXwdLeh3:-KMzVfNvWN1m_TkCftZd:-883928159) ([raw file](https://github.com/robotlocomotion/drake/blob/638580f04e57c40658ac51081416ed2a79a8c0d2/drake/common/memory_buffer_utils.h#L24)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> > MemoryBufferToString
> 
> AFAICT this is only ever called from the unit test.  Why is it in this PR?
> 
> It is difficult to evaluate its design without context about how it will be used -- is there a dev branch where its use is demonstrated?
</details>
I used it to debug this PR. Once debugged, I am no longer using it. But I thought it would be useful in the future. Should I not include it in this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/memory_buffer_utils.h, line 24 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KMzVPsI8d2kaXwdLeh3:-KMzWayyNFGOk64uqDAL:-1061866499) ([raw file](https://github.com/robotlocomotion/drake/blob/638580f04e57c40658ac51081416ed2a79a8c0d2/drake/common/memory_buffer_utils.h#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I used it to debug this PR. Once debugged, I am no longer using it. But I thought it would be useful in the future. Should I not include it in this PR?
</details>
Perhaps a separate PR at best.  I suspect that "log this binary data" is best expressed not in a standalone function, but either as a feature of our logging API, or an `operator<<` or `ToString` member method of a `ByteBuffer` class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/memory_buffer_utils.h, line 24 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KMzVPsI8d2kaXwdLeh3:-KMzXvavxkdL94rTOP7g:-938671445) ([raw file](https://github.com/robotlocomotion/drake/blob/638580f04e57c40658ac51081416ed2a79a8c0d2/drake/common/memory_buffer_utils.h#L24)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Perhaps a separate PR at best.  I suspect that "log this binary data" is best expressed not in a standalone function, but either as a feature of our logging API, or an `operator<<` or `ToString` member method of a `ByteBuffer` class.
</details>
Done. PTAL. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 8 files at r9.
Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 20 files at r5, 3 of 13 files at r7, 1 of 3 files at r8, 3 of 8 files at r9.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 61 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-IT_dJ0CcCu-qrEtx:-KN-IT_dJ0CcCu-qrEty:2119135046) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L61)):*
> ```objc
>    *
>    * @param[in] vector_interface A reference to the
>    * `drake::systems::VectorInterface<double>` object to convert into an LCM
> ```

BTW I'm pretty sure doxygen shows the types as part of the API, so repeating the typename of the argument here is both redundant and thus brittle with respect to API changes.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 63 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-I_hFufJneP5RJgQY:-KN-I_hFufJneP5RJgQZ:920448163) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L63)):*
> ```objc
>    * `drake::systems::VectorInterface<double>` object to convert into an LCM
>    * message. This reference must be valid throughout the duration of this
>    * method call.
> ```

It seems like this last sentence is true of all parameters ever, in which case it's just extra prose getting in the way of the important facets of the API.  (I see now that the other method on the API has the same issue, but I didn't notice it in that PR.)

Perhaps the intent is to clarify that the method does not alias and retain the reference beyond the method call?  In that case, perhaps say that more directly?  (For me, "not aliased by the method" is the default semantics so not worth mentioning, but I could certainly live with stating it explicitly at author's discretion.)

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 67 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Ioa5ZFk7wV2-xa0I:-KN-Ioa6o2igAIrmmsnI:-394982410) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L67)):*
> ```objc
>    * @param[in] channel The name of the channel on which to publish the LCM
>    * message. This reference must be valid throughout the duration of this
>    * method call.
> ```

Ditto on the last sentence.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 70 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Itsm_Cm-MmYmFOTU:-KN-Itsm_Cm-MmYmFOTV:-394982410) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L70)):*
> ```objc
>    *
>    * @param[in] lcm A pointer to the LCM subsystem. This pointer must not be
>    * null and must remain valid throughout the duration of this method call.
> ```

Ditto on the last sentence.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 10 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LCqAYJViQe2JSZ9i:-KN-LCqAYJViQe2JSZ9j:355372729) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L10)):*
> ```cpp
> 
> LcmPublisherSystem::LcmPublisherSystem(const std::string& channel,
>                       const LcmAndVectorInterfaceTranslator& translator,
> ```

Mis-indented?  `clang-format`?

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 29 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LJYNv1z46u34Cfl2:-KN-LJYNv1z46u34Cfl3:1287220335) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L29)):*
> ```cpp
>     new BasicVector<double>(translator_.get_vector_size()));
> 
>   std::unique_ptr<InputPort<double>> inputPort(
> ```

`input_port`

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 46 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LWT-SvkR5dmkg_ty:-KN-LWT-SvkR5dmkg_tz:355372729) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L46)):*
> ```cpp
> 
> void LcmPublisherSystem::EvalOutput(const Context<double>& context,
>                 SystemOutput<double>* output) const {
> ```

Mis-indented?  `clang-format`?

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 54 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LeHgSljRujyLt4OU:-KN-LeHgSljRujyLt4OV:-2007298510) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L54)):*
> ```cpp
> 
>   // Translates the input vector into an LCM message. The data is stored in
>   // the memory pointed to by variable buffer.
> ```

The second sentence doesn't make sense to me.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 28 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KPr8pOc31UDRWwqe:-KN-KPr8pOc31UDRWwqf:1653847797) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L28)):*
> ```objc
>    *
>    * @param[in] translator The translator that converts between LCM message
>    * objects and `drake::systems::VectorInterface` objects.
> ```

Document the lifetime (the translator is aliased).

---

*[drake/systems/lcm/lcm_publisher_system.h, line 51 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KhI34hEEdC9yL9by:-KN-KhI4x0QIpP1MOMVH:951003168) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L51)):*
> ```objc
> 
>   /**
>    * The output consists of a single port containing a `BasicVector<double>`.
> ```

Is this true?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 65 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KmZl8AUZXYigv657:-KN-KmZl8AUZXYigv658:1484089318) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L65)):*
> ```objc
>  private:
>   const int kNumInputPorts = 1;
>   const int kPortIndex = 0;
> ```

To the extent possible, it's always(?) better to put implementation details into the `.cc` file instead of the `.h` file.  These two integers don't seem to be needed here.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 71 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KyyhV8Sbt6MxPS97:-KN-KyyiaxgSHJWwO5nA:-213961217) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L71)):*
> ```objc
> 
>   // The translator that converts between LCM messages and
>   // drake::systems::BasicVector objects.
> ```

Is the `BasicVector` part here still accurate?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 75 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-L1oU_Y7ks7xZ8aRE:-KN-L1oU_Y7ks7xZ8aRF:-2081927661) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L75)):*
> ```objc
> 
>   // A pointer to the LCM subsystem.
>   ::lcm::LCM* lcm_;
> ```

BTW Consider `::lcm::LCM* const lcm_;` for clarity.

---

*[drake/systems/lcm/translator_between_lcmt_drake_signal.cc, line 58 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-K9HOjy3KSiS0ilYM:-KN-K9HOjy3KSiS0ilYN:-240903832) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/translator_between_lcmt_drake_signal.cc#L58)):*
> ```cpp
>   // Instantiates and initializes a LCM message capturing the state of
>   // parameter vector_interface.
>   drake::lcmt_drake_signal message;
> ```

BTW Consider documenting that `timestamp` is left unset on purpose.

---

*[drake/systems/lcm/translator_between_lcmt_drake_signal.cc, line 68 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-JodwwYIy5k5ydwjn:-KN-JodwwYIy5k5ydwjo:1351632836) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/translator_between_lcmt_drake_signal.cc#L68)):*
> ```cpp
>   for (int ii = 0; ii < message.dim; ++ii) {
>     message.val[ii] = values[ii];
>     message.coord[ii] = "Coord_" + std::to_string(ii);
> ```

This is a lot of string arithmetic to be doing for every message to send.  Is there any benefit or need to fill in these information-less coord names?

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 46 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-MauJPP1Dp7p9ZvcK:-KN-MauJPP1Dp7p9ZvcL:1306041998) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L46)):*
> ```cpp
>     drake::lcmt_drake_signal message_copy;
> 
>     message_mutex_.lock();
> ```

Use the RAII locker, like the prior PR.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 54 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Mjg0dyZKxRGotQ2J:-KN-Mjg0dyZKxRGotQ2K:231049113) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L54)):*
> ```cpp
> 
>  private:
>   void HandleMessage(const ::lcm::ReceiveBuffer *rbuf,
> ```

Whitespace `*` near type, not variable. `clang-format`?

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 55 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-MoTT3_BuHgmodB4A:-KN-MoTT3_BuHgmodB4B:1700997001) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L55)):*
> ```cpp
>  private:
>   void HandleMessage(const ::lcm::ReceiveBuffer *rbuf,
>                      const std::string &channel_name) {
> ```

Ditto.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 57 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Mqeuzh31fkfCPVn4:-KN-Mqeuzh31fkfCPVn5:-131383901) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L57)):*
> ```cpp
>                      const std::string &channel_name) {
>     if (channel_name_ == channel_name) {
>       message_mutex_.lock();
> ```

Use RAII locker.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 58 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Muw0-EDNv6PHy_PD:-KN-Muw0-EDNv6PHy_PE:-541143228) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L58)):*
> ```cpp
>     if (channel_name_ == channel_name) {
>       message_mutex_.lock();
>       received_message_.decode(rbuf->data, 0, rbuf->data_size);
> ```

I think `decode` can return an error, which this code is ignoring?  At a minimum, it should at least be explicit that it's ignoring the error.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 118 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NXWOrchKCyxrZyVx:-KN-NXWOrchKCyxrZyVy:-723547233) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L118)):*
> ```cpp
>     new FreestandingInputPort<double>(std::move(basic_vector)));
> 
>   context->SetInputPort(kPortNumber, std::move(input_port));
> ```

Isn't this an abuse of the System 2 APIs?  I thought the ideas was to set values on the context objects, not replace the context with different instances?

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 162 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NrShWipYgEuOh1Ru:-KN-NrShWipYgEuOh1Rv:-1333608073) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L162)):*
> ```cpp
>       //
>       // Thus, we must conclude that the experiment succeeded.
>       if (values_match) done = true;
> ```

The phrasing of the paragraph of text above this ("contains the expected values") and this if-statement ("were the expected values found?") seem a bit at odds.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 8 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 61 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-IT_dJ0CcCu-qrEtx:-KN-kcjvT7webMeJptnF:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I'm pretty sure doxygen shows the types as part of the API, so repeating the typename of the argument here is both redundant and thus brittle with respect to API changes.
</details>
Done.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 63 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-I_hFufJneP5RJgQY:-KN-kpKVhhXqfsbrUI1v:1453212188) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L63)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It seems like this last sentence is true of all parameters ever, in which case it's just extra prose getting in the way of the important facets of the API.  (I see now that the other method on the API has the same issue, but I didn't notice it in that PR.)
> 
> Perhaps the intent is to clarify that the method does not alias and retain the reference beyond the method call?  In that case, perhaps say that more directly?  (For me, "not aliased by the method" is the default semantics so not worth mentioning, but I could certainly live with stating it explicitly at author's discretion.)
</details>
Done. OK for not mentioning lifetime expectations explicitly by assuming validity throughout method call.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 67 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Ioa5ZFk7wV2-xa0I:-KN-l3bb0I4BTFYkQQQj:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L67)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Ditto on the last sentence.
</details>
Done.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 70 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Itsm_Cm-MmYmFOTU:-KN-l4YUgG3rWAG0t5Vq:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_and_vector_interface_translator.h#L70)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Ditto on the last sentence.
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 10 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LCqAYJViQe2JSZ9i:-KN-lPbSLLAlvsh5jegk:123980880) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Mis-indented?  `clang-format`?
</details>
Done by running `clang-format`.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 29 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LJYNv1z46u34Cfl2:-KN-lcVXCiR-uOjpAJac:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `input_port`
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 46 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LWT-SvkR5dmkg_ty:-KN-ljmvx9NdL7O8scyL:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Mis-indented?  `clang-format`?
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 54 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-LeHgSljRujyLt4OU:-KN-luGtIchckFu7Xk1G:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.cc#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The second sentence doesn't make sense to me.
</details>
Fixed. Thanks.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 28 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KPr8pOc31UDRWwqe:-KN-m7iwCOkZLj_XfDhJ:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Document the lifetime (the translator is aliased).
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 51 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KhI34hEEdC9yL9by:-KN-mf_TpdBJnBgeKTAq:-1072264579) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L51)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Is this true?
</details>
Oops. Fixed. This system has no output ports.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 65 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KmZl8AUZXYigv657:-KN-myfS1sl_wo_6JpVe:950139883) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> To the extent possible, it's always(?) better to put implementation details into the `.cc` file instead of the `.h` file.  These two integers don't seem to be needed here.
</details>
Done. Moved into an anonymous namespace in the `.cc` file.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 71 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-KyyhV8Sbt6MxPS97:-KN-n4Q59HGvuJoE9Lb4:-818947955) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/lcm_publisher_system.h#L71)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Is the `BasicVector` part here still accurate?
</details>
Oops. Fixed. It's `VectorInterface` objects.

---

*[drake/systems/lcm/translator_between_lcmt_drake_signal.cc, line 68 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-JodwwYIy5k5ydwjn:-KN-nGshCTqrUIZcbCUe:-1743714563) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/translator_between_lcmt_drake_signal.cc#L68)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This is a lot of string arithmetic to be doing for every message to send.  Is there any benefit or need to fill in these information-less coord names?
</details>
I removed it since I'm not sure it's required. The definition of `lcmt_drake_signal` does not specify whether the `coord` vector must contain valid value:

https://github.com/RobotLocomotion/drake/blob/master/drake/lcmtypes/lcmt_drake_signal.lcm

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 46 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-MauJPP1Dp7p9ZvcK:-KN-nzK69a0pSsoNvew0:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Use the RAII locker, like the prior PR.
</details>
Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 54 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Mjg0dyZKxRGotQ2J:-KN-oGJWZzVi2UaxN1VG:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Whitespace `*` near type, not variable. `clang-format`?
</details>
Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 55 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-MoTT3_BuHgmodB4A:-KN-oJErX76rYZumqrLp:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Ditto.
</details>
Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 57 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Mqeuzh31fkfCPVn4:-KN-oKF9VoDyXk4qDoX3:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L57)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Use RAII locker.
</details>
Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 58 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-Muw0-EDNv6PHy_PD:-KN-p47bJ-F6wr62oyG0:-367069225) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L58)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think `decode` can return an error, which this code is ignoring?  At a minimum, it should at least be explicit that it's ignoring the error.
</details>
Yes, that is true. I added a comment explaining why the error can be ignored in this context.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 118 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NXWOrchKCyxrZyVx:-KN-vRpqZ7GH-rlO0Nts:-1063997524) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Isn't this an abuse of the System 2 APIs?  I thought the ideas was to set values on the context objects, not replace the context with different instances?
</details>
I can see how it may be an abuse of the API since the input port is already created by `LcmPublisherSystem::CreateDefaultContext()`. However, looking at the [API of drake::systems::Context](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html), I don't see any other way to initialize the values in the context's input port. Method [get_vector_input](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html#af0477e781c2835067fe7955fa0495d7c) returns a pointer to a `const VectorInput`, which disallows modification of its vector. Thus, I believe my only option is to call [SetInputPort()](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html#ad51f36f03cd48e7346384719b1a90757).

I updated the comment above to include the above explanation.

@david-german-tri, is the above reasoning correct?

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 162 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NrShWipYgEuOh1Ru:-KN-vtpK9QgzweqmM8Em:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The phrasing of the paragraph of text above this ("contains the expected values") and this if-statement ("were the expected values found?") seem a bit at odds.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2781/ and is thus ready for another round of review.

jwnimmer-tri commented:
LGTM ready for platform review.

---

Reviewed 5 of 6 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN232kXqbMDtdvDkKKQ:-497205564)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 118 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NXWOrchKCyxrZyVx:-KN235wmF5GZUy9S9v5o:-1670165377) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L118)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I can see how it may be an abuse of the API since the input port is already created by `LcmPublisherSystem::CreateDefaultContext()`. However, looking at the [API of drake::systems::Context](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html), I don't see any other way to initialize the values in the context's input port. Method [get_vector_input](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html#af0477e781c2835067fe7955fa0495d7c) returns a pointer to a `const VectorInput`, which disallows modification of its vector. Thus, I believe my only option is to call [SetInputPort()](http://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_context.html#ad51f36f03cd48e7346384719b1a90757).
> 
> I updated the comment above to include the above explanation.
> 
> @david-german-tri, is the above reasoning correct?
</details>
I see.  I'll mark this blocking until @david-german-tri can weigh in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Since today is Tuesday, +@david-german-tri for platform review.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN2F7DhE5DIZ11B92ND:-746278951)*
<!-- Sent from Reviewable.io -->

liangfok commented:
In this PR, `LcmPublisherSystem` publishes an LCM message within method `EvalOutput()`. This is contrary to what @sherm1 desired based on his comment [here](https://github.com/RobotLocomotion/drake/pull/2729#issuecomment-230664855):

> If the System has no OutputPorts then there should be no need to call its EvalOutput() method. Instead it should have an LcmOutputSystem-specific method that is invoked to obtain the LCM message, something like EvalLcmMessage().

Since, `SystemInterface` currently does not provide an `EvalLcmMessage()` or `EvalMiddlewareCommunication()`, this PR has no choice but to implement the publishing of LCM message in `EvalOutput()`.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN2g4JRUw0leDNF6AAE:1093003022)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 68 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2eO2zuNYQfu2m_qJw:-KN2eO2zuNYQfu2m_qJx:1803567937) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_and_vector_interface_translator.h#L68)):*
> ```objc
>    * `nullptr`.
>    */
>   virtual void TranslateAndSendVectorInterfaceToLCM(
> ```

BTW, alternative name idea: `PublishVectorInterfaceToLCM`

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 30 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2htPWXsrIcXkwxB7y:-KN2htPWXsrIcXkwxB7z:-547750925) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.cc#L30)):*
> ```cpp
>       new BasicVector<double>(translator_.get_vector_size()));
> 
>   std::unique_ptr<InputPort<double>> input_port(
> ```

As far as I can tell, creating this FreestandingInputPort isn't adding any value.  Unless you have a specific need for a default input (which should rarely or never be true), it's better just to call `SetNumInputPorts`, and leave the ports themselves `nullptr`.  That way you get a hard failure if the calling code fails to wire up the system correctly.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 46 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2g_jgMV2BWs6UpkgW:-KN2g_jhrMoGajjDr2JG:-1128372029) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.cc#L46)):*
> ```cpp
> }
> 
> void LcmPublisherSystem::EvalOutput(const Context<double>& context,
> ```

Please add a TODO citing #2836.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 16 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2hCBChg4zy3O8ZPE2:-KN2hCBDVn7b3Lj_SpsX:1954000483) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.h#L16)):*
> ```objc
> 
> /**
>  * Receives `BasicVector<double>` values from a `SystemInterface<double>`'s
> ```

s/BasicVector/VectorInterface, here and below

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 118 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NXWOrchKCyxrZyVx:-KN2iJTVzP4zAfyI2dgu:-1041769864) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L118)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I see.  I'll mark this blocking until @david-german-tri can weigh in.
</details>
Yes, this is correct use of the API.  Input ports are inserted into the context by the code that builds the graph of Systems.  However, allocating an input port in `CreateDefaultContext` is unnecessary and confusing - see comment above.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 106 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2iplCD698a8ec2H-b:-KN2iplDRAvLN_bsy8Bz:306715751) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/test/lcm_publisher_system_test.cc#L106)):*
> ```cpp
> 
>   {
>     VectorX<double> vector_value;
> ```

BTW, this would be a good place to use `VectorInterface::get_mutable_value`.  It's not performance critical in this test, but why not set a good example for future readers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
BTW, I was just looking at that. We need something better than a side effect of EvalOutput as a way to get a System to perform work that is invisible to the System diagram abstractions. In general a System has multiple output ports with different sampling rates -- in this case there are zero output ports so I wouldn't expect a solver to call EvalOutputPort at all. A System block itself has a sampling rate also and we should probably be triggering its side-effect activity off the block sample rather than off a (non-existent) output port sample. However that capability does not exist in System 2 at the moment so shouldn't be an issue for this PR to deal with.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN2iQsbKXK_7PV9boSX:-1445369853)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I believe I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/lcm/lcm_and_vector_interface_translator.h, line 68 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2eO2zuNYQfu2m_qJw:-KN2sIqy7d1haqAe2iak:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_and_vector_interface_translator.h#L68)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, alternative name idea: `PublishVectorInterfaceToLCM`
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 30 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2htPWXsrIcXkwxB7y:-KN2n0a3-jxWfyoS6PDu:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.cc#L30)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> As far as I can tell, creating this FreestandingInputPort isn't adding any value.  Unless you have a specific need for a default input (which should rarely or never be true), it's better just to call `SetNumInputPorts`, and leave the ports themselves `nullptr`.  That way you get a hard failure if the calling code fails to wire up the system correctly.
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 46 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2g_jgMV2BWs6UpkgW:-KN2mvHpnb3lQvxvOoCW:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.cc#L46)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Please add a TODO citing #2836.
</details>
Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 16 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2hCBChg4zy3O8ZPE2:-KN2n9ZDuPeQkqu1TmZy:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/lcm_publisher_system.h#L16)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> s/BasicVector/VectorInterface, here and below
</details>
Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 118 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN-NXWOrchKCyxrZyVx:-KN2nbe6se_Z5AsLTHuB:-255290377) ([raw file](https://github.com/robotlocomotion/drake/blob/d42585bad6bda11d68469c63831246d6f4e5b9ff/drake/systems/lcm/test/lcm_publisher_system_test.cc#L118)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yes, this is correct use of the API.  Input ports are inserted into the context by the code that builds the graph of Systems.  However, allocating an input port in `CreateDefaultContext` is unnecessary and confusing - see comment above.
</details>
Done. (No change here, see change in `LcmPublisherSystem::CreateDefaultContext()`.)

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 106 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2iplCD698a8ec2H-b:-KN2uKIi_fB78UUIs9sY:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42a6c93b7e639db567cb8ce901dfbfe38048fb76/drake/systems/lcm/test/lcm_publisher_system_test.cc#L106)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, this would be a good place to use `VectorInterface::get_mutable_value`.  It's not performance critical in this test, but why not set a good example for future readers?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN2wZT1oqtLqM6-mIxK:963099758)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 110 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2z699Mk_CVAMsohD-:-KN2z69AU2OdPef-Eo7j:-1165801927) ([raw file](https://github.com/robotlocomotion/drake/blob/2c8b487db10d4c8b1b519cf1d735fa4baf0e23d5/drake/systems/lcm/test/lcm_publisher_system_test.cc#L110)):*
> ```cpp
>         vector_interface->get_mutable_value();
> 
>     vector_value.resize(kDim);
> ```

Resizing of a `get_mutable_value` is not allowed, per its documented API.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 110 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-KN2z699Mk_CVAMsohD-:-KN3-jEUwFpI8EnS6hNa:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/2c8b487db10d4c8b1b519cf1d735fa4baf0e23d5/drake/systems/lcm/test/lcm_publisher_system_test.cc#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Resizing of a `get_mutable_value` is not allowed, per its documented API.
</details>
Fixed. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@david-german-tri: This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2797/ and is thus ready for another round of review and/or merge.

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2821#-:-KN3MSw7W9qm2Q5-hJon:1417088721)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Looks like I may have broken continuous CI with this PR:

https://drake-jenkins.csail.mit.edu/job/linux-clang-continuous-debug/444/changes

Looking into it now...
